applications/OptimizationApplication/python_scripts/responses/response_function.py

```diff
@@ -1,27 +1,110 @@
 from abc import ABC, abstractmethod
+from typing import Union
 import KratosMultiphysics as Kratos
+import KratosMultiphysics.OptimizationApplication as KratosOA
 from KratosMultiphysics.OptimizationApplication.utilities.union_utilities import SupportedSensitivityFieldVariableTypes
 
 class ResponseFunction(ABC):
-    @classmethod
+    """Base response function.
+
+    This reponse function is the base response function. This is assumed to have following responsibilities.
+        1. CalculateValue for a fresh design. (@see CalculateValue)
+        2. CalculateSensitivity for a fresh design (@see CalculateSensitivity)
+
+    This control should only work on one model part. Hence, if multiple model parts required then,
+    a single model part should be created using Kratos.ModelPartOperationUtilities.
+    """
+
     @abstractmethod
-    def GetSensitivityFieldVariables(cls) -> 'list[SupportedSensitivityFieldVariableTypes]':
+    def Initialize(self) -> None:
+        """Initializes the response.
+
+        This method initializes the response. This is only called once in the whole optimization process.
+
+        """
         pass
 
-    def Initialize(self) -> None:
+    @abstractmethod
+    def Check(self) -> None:
+        """Checks the response.
+
+        This method checks the response. This is only called once in the whole optimization process.
+
+        """
         pass
 
+    @abstractmethod
     def Finalize(self) -> None:
+        """Finalizes the response.
+
+        This method finalizes the response. This is only called once in the whole optimization process.
+
+        """
         pass
 
     @abstractmethod
-    def Check(self) -> None:
+    def GetDependentPhysicalKratosVariables(self) -> 'list[SupportedSensitivityFieldVariableTypes]':
+        """Returns all the dependent physical kratos variables of the response.
+
+        This method should return all the dependent physical variables of the response to make sure
+        that sensitivities w.r.t. those variables are requested from the response. If sensitivities
+        w.r.t. other variables are required from this response, they are ASSUMED TO BE ZERO.
+
+        Please return all the dependent physical variables, eventhough the gradient computation is not yet implemented
+        for some to avoid future bugs.
+
+        Returns:
+            list[SupportedSensitivityFieldVariableTypes]: All dependent physical variables of the response.
+        """
         pass
 
     @abstractmethod
     def CalculateValue(self) -> float:
+        """Calculates the response value.
+
+        This method should always calculate the response value assuming the domain has changed.
+
+        Returns:
+            float: Calculated response value.
+        """
+        pass
+
+    @abstractmethod
+    def CalculateGradient(self, physical_variable_collective_expressions: 'dict[SupportedSensitivityFieldVariableTypes, KratosOA.ContainerExpression.CollectiveExpressions]') -> None:
```

Comment:
To facilitate and simplify the implementation of optimization algorithms, I would make the design collective_expressions as input to this function. Similar to the dependency check in the MapGradient function of Master control, we check if there is a dependency and if yes then compute sens for the found sensitivity types

Reply:
see #11077